src/store/index.js

```diff
@@ -11,28 +11,86 @@ const initialState = {
     },
     isKeyboardEventsEnabled: false,
     isMouseEventsEnabled: false,
+    trackedEvents: {
+        isActive: false,
+        events: [],
+    },
 };
 
-const createStore = (instance, reducer, state) => {
+const createStore = (instance, reducer) => {
     const listeners = [];
 
     const getters = {
         isWidgetOpened: (overlayID) =>
             instance.store.state.overlay.isOpen && instance.store.state.overlay.widgetOpened.includes(overlayID),
     };
 
-    const dispatch = (action) => {
-        instance.store.state = reducer(instance.store.state, action);
-        listeners.forEach(({cb}) => {
-            cb();
+    const hasChanged = (changedKeys, keyPath) => {
+        const keys = keyPath.split('.');
+        return changedKeys.some((changedKey) => {
+            const changedKeyParts = changedKey.split('.');
+            return keys.every((key, index) => key === changedKeyParts[index]);
+        });
+    };
+
+    const findChangedKeys = (newState, oldState, path = []) => {
+        let changedKeys = [];
+
+        const isObject = (obj) => obj && typeof obj === 'object';
+
+        for (const key in newState) {
+            const fullPath = [...path, key].join('.');
+
+            if (!Object.prototype.hasOwnProperty.call(oldState, key)) {
+                changedKeys.push(fullPath);
+            } else if (isObject(newState[key]) && isObject(oldState[key])) {
+                if (Array.isArray(newState[key]) && Array.isArray(oldState[key])) {
+                    if (newState[key].length !== oldState[key].length) {
+                        changedKeys.push(fullPath);
+                    } else {
+                        let arrayHasChanged = false;
+                        for (let i = 0; i < newState[key].length; i++) {
+                            if (newState[key][i] !== oldState[key][i]) {
+                                arrayHasChanged = true;
+                            }
+                        }
+                        if (arrayHasChanged) {
+                            changedKeys.push(fullPath);
+                        }
+                    }
+                } else {
+                    changedKeys = changedKeys.concat(findChangedKeys(newState[key], oldState[key], [...path, key]));
+                }
+            } else if (newState[key] !== oldState[key]) {
+                changedKeys.push(fullPath);
+            }
+        }
+
+        return changedKeys;
+    };
```

Comment:
It would be nice for readibility to refactor this function in order to avoid such deep nesting of conditions & loops. For example by extracting to a function the array check ?

```suggestion
    const arraysAreEqual = (firstArray, secondArray) => {
        if (firstArray.length !== secondArray.length) {
            return true;
        }
        for (let i = 0; i < firstArray.length; i++) {
            if (firstArray[i] !== secondArray[i]) {
                return true;
            }
        }
        return false;
    }
    const findChangedKeys = (newState, oldState, path = []) => {
        let changedKeys = [];

        const isObject = (obj) => obj && typeof obj === 'object';

        for (const key in newState) {
            const fullPath = [...path, key].join('.');

            if (!Object.prototype.hasOwnProperty.call(oldState, key)) {
                changedKeys.push(fullPath);
            } else if (isObject(newState[key]) && isObject(oldState[key])) {
                if (Array.isArray(newState[key]) && Array.isArray(oldState[key]) && !arraysAreEqual(newState[key], oldState[key])) {
                    changedKeys.push(fullPath);
                } else {
                    changedKeys = changedKeys.concat(findChangedKeys(newState[key], oldState[key], [...path, key]));
                }
            } else if (newState[key] !== oldState[key]) {
                changedKeys.push(fullPath);
            }
        }

        return changedKeys;
    };
```

But there's probably ways to reduce even more the nesting.